### PR TITLE
chore: Execute the whole engine natively in the cli

### PR DESCRIPTION
**Pull request by Finistere**
# Description

Please include a summary of the change and which issue is fixed.
Please also include relevant motivation and context.

# Type of change

- [ ] 💔 Breaking
- [ ] 🚀 Feature
- [ ] 🐛 Fix
- [ ] 🛠️ Tooling
- [ ] 🔨 Refactoring
- [ ] 🧪 Test
- [ ] 📦 Dependency
- [ ] 📖 Requires documentation update

### CODE REVIEW

File: cli/crates/server/src/parser.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
+use std::collections::{HashMap, HashSet};
```
Comment (Finistere) [inline cli/crates/server/src/parser.rs:1]: pretty much `parser-wasm` from api

File: engine/crates/gateway-core/src/serving.rs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+// with a GET request, we try to build the query from those parameters.
+pub const QUERY_REQUEST_PARAMETER: &str = "query";
```
Comment (Finistere) [inline engine/crates/gateway-core/src/serving.rs:2]: just to ensure some consistency between cloud and local

File: cli/crates/gateway/src/context.rs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+use http::HeaderMap;
+use tokio::{sync::Mutex, task::JoinHandle};
+
+pub struct Context {
+    pub(crate) ray_id: String,
+    pub(crate) x_api_key_header: Option<String>,
+    pub(crate) authorization_header: Option<String>,
+    pub(crate) headers: HeaderMap,
+    // TODO: or use a queue?
+    pub(crate) join_handles: Mutex<Vec<JoinHandle<()>>>,
```
Comment (obmarg) [inline cli/crates/gateway/src/context.rs:11]: Unbounded channels _are_ easier but 🤷

File: cli/crates/gateway/src/context.rs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+impl gateway_core::RequestContext for Context {
+    fn ray_id(&self) -> &str {
+        &self.ray_id
+    }
+
+    async fn wait_until(&self, fut: BoxFuture<'static, ()>) {
+        // FIXME: should be propagated.
+        let handle = tokio::runtime::Handle::current();
+        self.join_handles.lock().await.push(handle.spawn(fut));
+    }
```
Comment (obmarg) [inline cli/crates/gateway/src/context.rs:24]: I'm not sure that this should spawn.  Don't think it was the behaviour in the existing gateway anyway - pretty sure it waits until the request is finished before running the `wait_until` futures.

File: cli/crates/gateway/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,58 @@
+use gateway_core::CacheConfig;
+// Only included to force local feature
```
Comment (obmarg) [inline cli/crates/gateway/src/lib.rs:2]: I don't get this comment

File: cli/crates/gateway/src/serving.rs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            .get(X_API_KEY_HEADER)
+            .and_then(|value| value.to_str().ok().map(ToString::to_string))
+            .or_else(|| params.get(X_API_KEY_HEADER).cloned()),
+        authorization_header: headers
+            .get(AUTHORIZATION_HEADER)
+            .and_then(|value| value.to_str().ok().map(ToString::to_string))
+            .or_else(|| params.get(AUTHORIZATION_HEADER).cloned()),
+        headers,
+        join_handles: Mutex::default(),
+    });
```
Comment (obmarg) [inline cli/crates/gateway/src/serving.rs:53]: Could probably put this in a function somewhere

File: engine/crates/gateway-core/src/lib.rs [hunk 7]
```diff
@@ -1,151 +1,127 @@
...
-
-    async fn authorize_admin_request(&self, _request: &async_graphql::Request) -> Result<(), AdminAuthError>;
+// A bit tedious but this allows the caller to have an easier time with its error, response and
+// context types.
+pub struct Gateway<Executor, Cache, Context, Response, Error> {
+    executor: Arc<Executor>,
+    cache: Arc<Cache>,
+    cache_config: CacheConfig,
+    authorizer: Box<dyn Authorizer<Context = Context>>,
+    _phantom: PhantomData<fn(Context) -> (Response, Error)>,
```
Comment (obmarg) [inline engine/crates/gateway-core/src/lib.rs:41]: Context appears on that `dyn Authorizor` so you probably don't need it in the phantom

```suggestion
    authorizer: Box<dyn Authorizer<Context = Context>>,
    _phantom: PhantomData<fn() -> (Response, Error)>,
```

File: engine/crates/gateway-core/src/lib.rs [hunk 8]
```diff
@@ -1,151 +1,127 @@
...
 
-    async fn authorize_request(&self, _request: &engine::Request) -> Result<ExecutionAuth, AuthError>;
+impl<Executor, Cache, Context, Response, Error> Gateway<Executor, Cache, Context, Response, Error>
+where
+    Error: From<Error> + std::error::Error + Send + 'static,
+    Context: RequestContext,
+    Response: self::Response<Error = Error>,
+    Cache: self::Cache<Value = engine::Response> + 'static,
+    Executor: self::Executor<Context = Context, Error = Error, Response = Response> + 'static,
+{
```
Comment (obmarg) [inline engine/crates/gateway-core/src/lib.rs:51]: `Response` & `Error` come from the `Executor` so I think you could ditch them as generic params:

```suggestion
impl<Executor, Cache, Contextr> Gateway<Executor, Cache, Context>
where
    Executor: self::Executor<Context = Context> + 'static,
    Executor::Response: self::Response<Error = Executor::Error>,
    Executor::Error: From<Error> + std::error::Error + Send + 'static,
    Context: RequestContext,
    Cache: self::Cache<Value = engine::Response> + 'static,
{
```

**Review by obmarg — APPROVED** [review]

File: engine/crates/gateway-core/src/lib.rs (hunk 8 above)
Comment (Finistere) [reply engine/crates/gateway-core/src/lib.rs]: good catch!

**Pull request merged**